feat: Integrate discussion config API

omerhabib26 opened the pull request:
### Description:

- Integrate course discussion config Api
- Allow user to post based on discussion config
- Update test cases

fix: [LEARNER-10583](https://2u-internal.atlassian.net/browse/LEARNER-10583)


| Dark Mode  | Light Mode       | 
|-------------|----------------|
| <img width="341" alt="DarkMode" src="https://github.com/user-attachments/assets/e5978419-f4c5-4ba3-b73f-5da6ef275bd9" /> | <img width="343" alt="Light Mode" src="https://github.com/user-attachments/assets/42b5b968-c118-47a9-b6b2-2016afc57bff" />|

omerhabib26 committed:
feat: Integrate discussion config API

- Integrate course discussion config Api
- Allow user to post based on discussion config

fix: LEARNER-10583

omerhabib26 committed:
fix: update test cases

farhan-arshad-dev reviewed (changes requested):
Android design still needs minor design improvements to achieve parity with iOS, for both light and dark modes.

| Android  | iOS |
| ------------- | ------------- |
| ![image](https://github.com/user-attachments/assets/e1b71d8a-1f50-49ec-9bb7-d0f33f3185c5) | ![image](https://github.com/user-attachments/assets/7e4b1f35-c557-41f6-bd88-f5fe9bceb573)  |

farhan-arshad-dev commented on discussion/src/main/java/org/openedx/discussion/data/model/response/DiscussionConfig.kt:
```diff
@@ -0,0 +1,103 @@
+                DomainReason(it.code, it.label)
+            },
+            postCloseReasons = postCloseReasons.map {
+                DomainReason(it.code, it.label)
```
[Optional]: IMO, it's better to create only the properties that are actually used in the code. We can still retain the full set of properties in the data model for completeness, but avoid defining unused ones in the implementation.

farhan-arshad-dev commented on discussion/src/main/java/org/openedx/discussion/presentation/threads/DiscussionThreadsViewModel.kt:
```diff
@@ -108,6 +109,8 @@ class DiscussionThreadsViewModel(
         viewModelScope.launch {
             try {
+                val discussionConfig = interactor.getCourseDiscussionConfig(courseId)
+                isPostingEnabled = discussionConfig.isPostingEnabled
```
It's better to call the endpoint & save the response to reuse the config. No need to call the endpoint multiple times. OR can pass via arguments.

farhan-arshad-dev commented on discussion/src/main/java/org/openedx/discussion/presentation/threads/DiscussionThreadsUIState.kt:
```diff
@@ -1,8 +1,10 @@
+    ) : DiscussionThreadsUIState()
 
     object Loading : DiscussionThreadsUIState()
 }
```
EOF is missing.

farhan-arshad-dev commented on discussion/src/main/java/org/openedx/discussion/presentation/topics/DiscussionTopicsViewModel.kt:
```diff
@@ -46,9 +46,13 @@ class DiscussionTopicsViewModel(
     private fun getCourseTopic() {
         viewModelScope.launch {
             try {
+                val discussionConfig = interactor.getCourseDiscussionConfig(courseId)
```
It's better to handle the error case.

omerhabib26 replied:
at the moment, we are only using this for a single flag, so need to handle it further.

omerhabib26 committed:
fix: Address PR comment

farhan-arshad-dev commented on discussion/src/main/java/org/openedx/discussion/presentation/search/DiscussionSearchThreadFragment.kt:
```diff
@@ -99,6 +99,7 @@ class DiscussionSearchThreadFragment : Fragment() {
         savedInstanceState: Bundle?
     ) = ComposeView(requireContext()).apply {
         setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+        val isPostingEnabled = requireArguments().getBoolean(ARG_IS_POSTING_ENABLED, true)
```
requires auto-format.

farhan-arshad-dev commented on discussion/src/main/java/org/openedx/discussion/presentation/search/DiscussionSearchThreadFragment.kt:
```diff
@@ -139,12 +141,16 @@ class DiscussionSearchThreadFragment : Fragment() {
 
```
extra line can be removed.

farhan-arshad-dev commented on discussion/src/main/java/org/openedx/discussion/presentation/comments/DiscussionCommentsFragment.kt:
```diff
@@ -119,6 +119,7 @@ class DiscussionCommentsFragment : Fragment() {
         savedInstanceState: Bundle?
     ) = ComposeView(requireContext()).apply {
         setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+        val isPostingEnabled = requireArguments().getBoolean(ARG_IS_POSTING_ENABLED, true)
```
We need to follow the consistent approach by getting the Arguments via the ViewModel. same goes for other screens

omerhabib26 committed:
fix: Address PR comment - 1

- Cached discussion config response for 1 hour
- Move isPostingEnabled to viewModel
- Update test cases accordingly
